Source/ASVideoPlayerNode.mm

```diff
@@ -254,61 +253,62 @@ - (void)didEnterPreloadState
-    _cachedControls = [[NSMutableDictionary alloc] init];
-  }
-
-  for (id object in _neededDefaultControls) {
-    ASVideoPlayerNodeControlType type = (ASVideoPlayerNodeControlType)[object integerValue];
-    switch (type) {
-      case ASVideoPlayerNodeControlTypePlaybackButton:
-        [self createPlaybackButton];
-        break;
-      case ASVideoPlayerNodeControlTypeElapsedText:
-        [self createElapsedTextField];
-        break;
-      case ASVideoPlayerNodeControlTypeDurationText:
-        [self createDurationTextField];
-        break;
-      case ASVideoPlayerNodeControlTypeScrubber:
-        [self createScrubber];
-        break;
-      case ASVideoPlayerNodeControlTypeFullScreenButton:
-        [self createFullScreenButton];
-        break;
-      case ASVideoPlayerNodeControlTypeFlexGrowSpacer:
-        [self createControlFlexGrowSpacer];
-        break;
-      default:
-        break;
+    if (_neededDefaultControls == nil) {
+      _neededDefaultControls = [self createDefaultControlElementArray];
+    }
+
+    if (_cachedControls == nil) {
+      _cachedControls = [[NSMutableDictionary alloc] init];
     }
-  }
 
-  if (_delegateFlags.delegateCustomControls && _delegateFlags.delegateLayoutSpecForControls) {
-    NSDictionary *customControls = [_delegate videoPlayerNodeCustomControls:self];
-    for (id key in customControls) {
-      id node = customControls[key];
-      if (![node isKindOfClass:[ASDisplayNode class]]) {
-        continue;
+    for (id object in _neededDefaultControls) {
+      ASVideoPlayerNodeControlType type = (ASVideoPlayerNodeControlType)[object integerValue];
+      switch (type) {
+        case ASVideoPlayerNodeControlTypePlaybackButton:
+          [self createPlaybackButton];
+          break;
+        case ASVideoPlayerNodeControlTypeElapsedText:
+          [self createElapsedTextField];
+          break;
+        case ASVideoPlayerNodeControlTypeDurationText:
+          [self createDurationTextField];
+          break;
+        case ASVideoPlayerNodeControlTypeScrubber:
+          [self createScrubber];
+          break;
+        case ASVideoPlayerNodeControlTypeFullScreenButton:
+          [self createFullScreenButton];
+          break;
+        case ASVideoPlayerNodeControlTypeFlexGrowSpacer:
+          [self createControlFlexGrowSpacer];
+          break;
+        default:
+          break;
       }
+    }
+
+    if (_delegateFlags.delegateCustomControls && _delegateFlags.delegateLayoutSpecForControls) {
+      NSDictionary *customControls = [_delegate videoPlayerNodeCustomControls:self];
+      for (id key in customControls) {
+        id node = customControls[key];
+        if (![node isKindOfClass:[ASDisplayNode class]]) {
+          continue;
+        }
 
-      [self addSubnode:node];
-      [_cachedControls setObject:node forKey:key];
+        [self addSubnode:node];
+        [_cachedControls setObject:node forKey:key];
+      }
     }
   }
 
   ASPerformBlockOnMainThread(^{
```

Comment:
Hah, good catch here. Since `ASPerformBlockOnMainThread` can run right away if already on main, the lock shouldn't be held when this line is reached.